### ISSUE TITLE
MutinyEmitter: No subscriber found for the channel X

### ISSUE THREAD

**Issue by dargmuesli** [smallrye/smallrye-reactive-messaging#839]
I updated to Quarkus 1.9 (from 1.6) and swapped usages of `Emitter` with `MutinyEmitter` to be able to handle Unis directly. Now my tests fail saying `error: java.lang.IllegalStateException: SRMSG00027: No subscriber found for the channel send-mail` for a declaration like this:
```java
@Inject
@Channel("send-mail")
MutinyEmitter<String> mailSendEmitter;
```

Somewhere below I only changed
`.onItem().transformToUni((mailMessage) -> Uni.createFrom().completionStage(mailSendEmitter.send(mailMessage)))`
to
`.onItem().transformToUni((mailMessage) -> mailSendEmitter.send(mailMessage))`

I haven't changed the `application.yaml`.

Do you have any idea why that could be?

**Comment by cescoffier** [comment]
That might be a bug. Does that only happen with `UniEmitter`?

**Comment by dargmuesli** [comment]
It happens with the `MutinyEmitter`. What other Emitter could I test? I'm fairly new to Quarkus / smallrye, but work on an existing codebase.

**Comment by cescoffier** [comment]
Sorry, `MutinyEmitter`. I was wondering if it happened with `Emitter`. 
`MutinyEmitter` has been integrated fairly recently, so we may have forgotten something. @kenfinnigan ?

**Comment by dargmuesli** [comment]
Ok, `Emitter` is where we come from, everything worked with that. The mentioned error does not appear, if a method is declared with `@Incoming("send-mail")`. Only when no such method exists and despite an outbound connection being defined like this:

```yaml
mp:
  messaging:
    outgoing:
      send-mail:
        connector: smallrye-amqp
        address: mailer
```

**Comment by cescoffier** [comment]
Can you provide a reproducer? I will have a look.

**Comment by kenfinnigan** [comment]
I think it's because there's no subscription to the returned Uni.

Can you try:

`.onItem().transformToUni((mailMessage) -> mailSendEmitter.send(mailMessage).subscribeWith(x -> {}))`

**Comment by cescoffier** [comment]
Oh right, thanks @kenfinnigan for spotting this. Without a subscription, the uni won't do anything!

**Comment by cescoffier** [comment]
@kenfinnigan Didn't we add something like `anfAwait` and `andForget` ?

**Comment by kenfinnigan** [comment]
Right, the alternative would be to call `mailSendEmitter.sendAndForget()`

**Comment by dargmuesli** [comment]
> Can you try:
>
> `.onItem().transformToUni((mailMessage) -> mailSendEmitter.send(mailMessage).subscribeWith(x -> {}))`

I guess you mean `.subscribe().with(x -> {})`? That and the suggested `.sendAndForget()` both return a cancellable. I'm a little confused what that means, as I'm fairly new to reactive messaging. I'm not sure about that `AndForget` too as I want to know when the email was sent? I thought using the `Uni<Void>` returned by `MutinyEmitter.send` with an `onItem()` is what I want :see_no_evil: 

I can surely try to create a reproducer, but that'll take some time.

**Comment by kenfinnigan** [comment]
It might be what you want, but you do need to subscribe to it for it to send.

**Comment by cescoffier** [comment]
@dargmuesli There is one essential difference between a `CompletionStage` and a `Uni.` `CompletionStages` are eager. So when using `CompletionStage<Void> cs = emitter.send(...)`, the message is sent, and the returned `CompletionStage` is a placeholder for the eventual outcome. `Unis,` on the other side, are lazy. `Uni<Void> uni = mutinyEmiter.send(...)` does not do anything until there is a _subscriber_ _expressing_ its interest in the action.

So for the `MutinyEmitter` to send the message, you need a subscriber. `onItem()` is not a subscriber; it's just another `Uni` (also lazy). You need a final subscriber. There are three ways to achieve this:

1. You can add a subscriber:

```java
emitter.send(...)
   .subscribe().with(success -> { ... }, failure -> { ... });
```

You mentioned the returned `Cancellable.` It's a reference on the computation you can use to _cancel_ it if you want to. Most of the time, you can ignore this (can be useful for long but interruptable tasks)

2. Use the "andForget` variant - subscribing under the hood but discarding the outcome:

```java
emitter.sendAndForget(...)
```

3. In Quarkus, HTTP endpoints, as well as some other parts of the API, can return `Uni,` and it's the _framework_ that subscribe and handle the outcome:

```
@Path("/")
@GET
public Uni<Void> send() {

   return emitter.send(...);

}
```

If you give us a bit more detail about where this is used, we can discuss the most suited approach.

**Comment by vankofalcon** [comment]
@cescoffier @kenfinnigan I'm still getting the same error when using
```
emitter.send(...)
   .subscribe().with(success -> { ... }, failure -> { ... });
```
that's in Quarkus 1.9.2
I'm using the smallrye-kafka as a connector for that channel
Is there a workaround or we have to wait for a fix?

**Comment by cescoffier** [comment]
is you emitter correctly connected to the right channel? Please provide a reproducer because if it's connected, that should not be a problem.

**Comment by vankofalcon** [comment]
`
@ApplicationScoped
public class MyManager {

@Inject
@Channel("channel-out")
MutinyEmitter<Obj> emitter;

public Uni<Obj> run() {
        return Uni.createFrom().item(new Obj())
                .onItem()
                .transformToUni(r -> {
                    emitter.send(r).subscribe().with((System.out::println, System.out::println));
                    return Uni.createFrom().item(r);
                });
    }
}    
    
@QuarkusMain
public class MainApp {
    public static void main(String... args) {
        Quarkus.run(MyApp.class, args);
    }

    public static class MyApp implements QuarkusApplication {

        @Inject
        MyManager mm;

        @Override
        public int run(String... args) throws Exception {
            mm.run().subscribe().with(System.out::println, System.out::println);
            Quarkus.waitForExit();
            return 0;
        }
    }
}


output: 

java.lang.IllegalStateException: SRMSG00027: No subscriber found for the channel channel-out
Obj(id=0, data=null, dateCreated=null)`

**Comment by kenfinnigan** [comment]
Is there any configuration for `channel-out` that connects it to Kafka or other provider?

**Comment by vankofalcon** [comment]
@kenfinnigan yes, it's a standard kafka cfg
```
mp.messaging.outgoing.channel-out.connector=smallrye-kafka
mp.messaging.outgoing.channel-out.topic=my-topic
mp.messaging.outgoing.channel-out.bootstrap.servers=servers-go-here
mp.messaging.outgoing.channel-out.value.serializer=ObjSerializer
```

**Comment by kenfinnigan** [comment]
Are there any log messages prior to "No subscriber" that indicates it wasn't able to connect to Kafka or "my-topic"?

**Comment by vankofalcon** [comment]
@kenfinnigan Only the normal kafka consumers connected output. Consumers are consuming messages. Everything with kafka works fine. I was using the non-reactive Emitter<T> before that and I've changed only that part to use the MutinyEmitter<T> api

**Comment by vankofalcon** [comment]
@kenfinnigan  Did you try running the sample code? Is it working for you?

**Comment by kenfinnigan** [comment]
I haven't run it.

Can you confirm that the code in `transformToUni` is executed?

**Comment by vankofalcon** [comment]
@kenfinnigan confirmed with a simlpe System.out.println :)

**Comment by vankofalcon** [comment]
I've updated the code snipped

**Comment by kenfinnigan** [comment]
With the updated snippet, there is no longer a `subscribe` on `emitter.send()`, which is required

**Comment by vankofalcon** [comment]
@kenfinnigan It doesn't matter how I write that code. In both cases I get the same error in the output. With that piece of code I still get the error `java.lang.IllegalStateException: SRMSG00027: No subscriber found for the channel channel-out`
```
public Uni<Obj> run() {
        return Uni.createFrom().item(new Obj())
                .onItem()
                .transformToUni(r -> {
                    System.out.println("transformToUni");
                    emitter.send(r).subscribe().with(System.out::println, System.out::println);
                    return Uni.createFrom().item(r);
                });
    }
}    
```

**Comment by cescoffier** [comment]
I will try to reproduce it.

**Comment by cescoffier** [comment]
@vankofalcon @kenfinnigan Found the issue. Mutiny emitters are not connected to channels managed by connectors. 
See https://github.com/smallrye/smallrye-reactive-messaging/pull/878 for a fix.

**Comment by cescoffier** [comment]
Fixed by #878.
